### PR TITLE
Add initial product data views screen

### PR DESCRIPTION
**Pull request by louwie17**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR is an initial part of the work done in this draft PR https://github.com/woocommerce/woocommerce/pull/50978 around adding a data views page for products that we can use to work on data forms.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Build this branch and create a new site with WooCommerce and the beta tester installed.
2. Hover over the **WooCommerce** menu and make sure it does **not** contain the **All Products** tab
3. Go to **Tools > WCA Test Helper > Features** and enable the **product-data-views** feature and refresh the page
4. Hover over the **WooCommerce** menu now, a **All Products** button should be present
5. Click on **All Products**
6. It should show a nearly white screen with **Please enabled Gutenberg for this feature**
7. Now install and enable Gutenberg
8. Refresh the new **All Products** page, it should now show **Initial Products Layout**
9. Enable the new **Product editor** under **WooCommerce > Settings > Advanced > Features**
10. Go to **Products > Add new** and make sure there are no errors ( the fact that it renders successfully without any crazy styling issues should be enough ).

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: packages/js/product-editor/src/lock-unlock.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/**
+ * External dependencies
+ */
+import { __dangerousOptInToUnstableAPIsOnlyForCoreModules } from '@wordpress/private-apis';
+
+export const { lock, unlock } =
+	__dangerousOptInToUnstableAPIsOnlyForCoreModules(
+		'I acknowledge private features are not for use in themes or plugins and doing so will break in the next version of WordPress.',
+		'@wordpress/edit-site'
+	);
```
Comment (louwie17) [inline packages/js/product-editor/src/lock-unlock.ts:10]: This only works within Gutenberg enabled, which is why I added the additional check in the `products.tsx` compoennt.

File: packages/js/product-editor/src/products-app/index.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * Internal dependencies
+ */
+import { unlock } from '../lock-unlock';
+
+const { RouterProvider } = unlock( routerPrivateApis );
+const { GlobalStylesProvider } = unlock( editorPrivateApis );
+
+function ProductsLayout() {
+	return <div>Initial Products Layout</div>;
```
Comment (louwie17) [inline packages/js/product-editor/src/products-app/index.tsx:21]: To be replaced by future work.

File: packages/js/product-editor/typings/index.d.ts [hunk 3]
```diff
@@ -25,9 +25,21 @@ declare module '@wordpress/core-data' {
...
+		isResolving: boolean;
+		hasResolved: boolean;
+	};
 }
 declare module '@wordpress/keyboard-shortcuts' {
-	function useShortcut(name: string, callback: (event: KeyboardEvent) => void): void;
+	function useShortcut(
+		name: string,
+		callback: ( event: KeyboardEvent ) => void
+	): void;
```
Comment (louwie17) [inline packages/js/product-editor/typings/index.d.ts:39]: Above changes are formatting.
I only added the below item for the router module.

File: plugins/woocommerce/src/Admin/Features/ProductDataViews/Init.php [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+			}
+
+			if ( ! $gutenberg_version ) {
+				$gutenberg_data    = get_file_data(
+					WP_PLUGIN_DIR . '/gutenberg/gutenberg.php',
+					array( 'Version' => 'Version' )
+				);
+				$gutenberg_version = $gutenberg_data['Version'];
+			}
+			return version_compare( $gutenberg_version, '19.0', '>=' );
```
Comment (louwie17) [inline plugins/woocommerce/src/Admin/Features/ProductDataViews/Init.php:69]: I used WP 6.6 and GB 19.0 for now, although DataViews has been enabled earlier.

**Review by gigitux — APPROVED** [review]
LGTM! Thanks for working on this! 🙇

**Pull request merged**